### ISSUE TITLE
HashTreeRoot Does Not Lead To The Same Result

### ISSUE THREAD

**Issue by nisdas** [prysmaticlabs/prysm#4526]
We currently have two methods to determine the hash tree root of the state. They are 
```
ssz.HashTreeRoot
```
and
```
stateutil.HashTreeRootState
```
however they lead to different roots for the below provided state.

[beacon_state.zip](https://github.com/prysmaticlabs/prysm/files/4054033/beacon_state.zip)

This was the root cause of our initial genesis failure in our testnet restart

**Comment by rauljordan** [comment]
Hey everyone, here is an analysis and discovery of the root cause explaining why it happened in the first place:

## TL;DR

`Go-SSZ.HTR` and Prysm's specific `stateutil.HTR` differ greatly in how they determine the sizes of certain fields for hashing. Go-SSZ looks directly at the protobuf field tags to determine sizes of fields while `stateutil.HTR` uses our `params.BeaconConfig()` to compute certain field sizes. Our genesis code in Prysm used `go-ssz.HTR(state)` while the entire rest of the repo uses `stateutil.HTR(state)` at runtime. Due to differences in config at compile time (such as not using the `--define ssz` flag), the `go-ssz.HTR` function resulted in a different root of the genesis state than what was expected by beacon proposers.

## Explanation 

Officially, the field `state.Eth1DataVotes` has `ssz-max:1024` as its size. When using `go-ssz.HTR`, the ssz code will directly read the protobuf tags to determine how it should hash the field. 

 In contrast, our `stateutil.HTR` uses a more specific approach of using `params.BeaconConfig().SlotsPerEth1VotingPeriod` as the max-size of `state.Eth1DataVotes`. This means if our config says 16 while the protobuf template has 1024, `go-ssz.HTR(state) != stateutil.HTR(state)`. If instead, we run the code above with `--define ssz=minimal`, both will match and pass any tests or required conditions. 

## Why This Broke the Testnet

For the testnet, we do not use minimal or mainnet config, but rather a customized `demo` config. Go-SSZ likely was not able to pick up on this change, but `stateutil.HTR` was, leading to discrepancies at runtime when attempting to use both in the same environment.

### Root Cause Discovery

This problem was reproducible in our e2e tests, after further investigation logging the results of `ssz.HashTreeRoot(genesis)` side by side with `stateutil.HashTreeRootState(genesis)` shows a few roots are different between their fields. Namely, `Eth1DataVotes`, `CurrentEpochAttestations`, and `PreviousEpochAttestations`. These fields are **fully** dependent on certain configuration parameters such as `SlotsPerEth1VotingPeriod`, so I had the suspicion it had to do with the config when building/running the beacon node.

If instead, we run `bazel test //endtoend:go_default_test --define ssz=minimal`, tests do pass, confirming the suspicion above. If this is not specified, we obtain the same error we saw in post-mortem number 7:

```
Could not compute state root: could not calculate state root at slot 0: could not process block: could not process block header: parent root 0x0e6c95db700881611523e282ec3fcaeeace681786aa47928c5903650165c6165 does not match the latest block header signing root in state 0x8b25301bdeba020d4226e949eae0ecefb8863e060b7844566d51e0a42ed6490e"
```
namely, the signing root of the block header does not match because the blocks had different state roots than what proposers expected. Proposers were running `stateutil.HTR(genesis)` while the original block's state root was determined with `go-ssz.HTR(genesis)`.

## Immediate, Simple Fix

Do not use `go-ssz.HTR(state)` anywhere in Prysm. This will fix this issue and eliminate it completely from our repo.

## Possible Long-Term Fix

We could revise our stateutil SSZ to parse template parameters of the state data structure instead of relying on parameter configurations. This would ensure consistent behavior when we use the `--define=minimal` or `mainnet` flags when running the system. If we want to be more flexible, however, relying on `params.BeaconConfig()` can be better. We would have to choose one over the other.

**Comment by rauljordan** [comment]
We can close this - we are no longer mixing go-ssz and stateutil HTR in Prysm
